plugins/device/transaction.go

```diff
@@ -1,52 +1,44 @@
 package device
 
 import (
-	"os/exec"
-	"strings"
-
 	"github.com/gluster/glusterd2/glusterd2/transaction"
 	deviceapi "github.com/gluster/glusterd2/plugins/device/api"
-
-	"github.com/pborman/uuid"
-	log "github.com/sirupsen/logrus"
+	"github.com/gluster/glusterd2/plugins/device/cmdexec"
 )
 
 func txnPrepareDevice(c transaction.TxnCtx) error {
-	var peerID uuid.UUID
-	var req deviceapi.AddDeviceReq
-	var deviceList []deviceapi.Info
-	if err := c.Get("peerid", peerID); err != nil {
-		c.Logger().WithError(err).Error("Failed transaction, cannot find peer-id")
+	var peerID string
+	if err := c.Get("peerid", &peerID); err != nil {
+		c.Logger().WithError(err).WithField("key", "peerid").Error("Failed to get key from transaction context")
 		return err
 	}
-	if err := c.Get("req", req); err != nil {
-		c.Logger().WithError(err).Error("Failed transaction, cannot find device-details")
+
+	var devices []string
+	if err := c.Get("devices", &devices); err != nil {
+		c.Logger().WithError(err).WithField("key", "req").Error("Failed to get key from transaction context")
 		return err
 	}
-	for _, name := range req.Devices {
+
+	var deviceList []deviceapi.Info
+	for _, name := range devices {
 		tempDevice := deviceapi.Info{
 			Name: name,
 		}
 		deviceList = append(deviceList, tempDevice)
 	}
-	for index, element := range deviceList {
-		pvcreateCmd := exec.Command("pvcreate", "--metadatasize=128M", "--dataalignment=256K", element.Name)
-		if err := pvcreateCmd.Run(); err != nil {
-			c.Logger().WithError(err).WithField("device", element.Name).Error("pvcreate failed for device")
-			deviceList[index].State = deviceapi.DeviceFailed
-			continue
-		}
-		vgcreateCmd := exec.Command("vgcreate", strings.Replace("vg"+element.Name, "/", "-", -1), element.Name)
-		if err := vgcreateCmd.Run(); err != nil {
-			c.Logger().WithError(err).WithField("device", element.Name).Error("vgcreate failed for device")
+
+	for index, device := range deviceList {
+		err := cmdexec.DeviceSetup(device.Name)
+		if err != nil {
```

Comment:
Log failure here and indicate in the message that you are setting the state to `DeviceFailed`.